### PR TITLE
[SYCL][Doc] Define buffer_location property for USM allocations

### PR DESCRIPTION
**Pull request by sherry-yuan**
This serves as a temporary solution to supporting usm buffer location, and it is the easiest to implement.
Align with changes in https://github.com/intel/llvm/pull/5634

A more complex workaround is to define a new malloc API that accept sycl::ext::oneapi::experimental::properties as its property argument of malloc.

The full solution is to define a malloc api that takes sycl::ext::oneapi::experimental::properties as property argument and returns annotated_ptr

### CODE REVIEW

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:1]: We are standardizing on a new template for extensions.  Can you use the template in #5663 for this extension?  In order to do that, you will need to decide:

* Is this a "oneapi" or an "intel" extension?
* Is this a "supported" or an "experimental" extension?

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:48]: I think I started a conversation about this someplace else too ... It would be better if this extension had some defined behavior when the device is not FPGA.  It seems like "buffer location" is a concept that only makes sense on FPGA-type devices.  In this case, it seems like it would make sense to say that the `buffer_location` is allowed when allocating memory for any device, but it only has an effect when the device is an FPGA.

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:92]: This namespace does not seem right.  I think it should be one of the following depending on whether this is a "oneapi" vs. an "intel" extension and on whether the extension is supported vs. experimental:

* `sycl::ext::oneapi::property::usm`
* `sycl::ext::intel::property::usm`
* `sycl::ext::oneapi::experimental::property::usm`
* `sycl::ext::intel::experimental::property::usm`

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:98]: I think the type of `location` should be `int`, not `uint64_t`.  This matches what we have in [sycl_ext_intel_buffer_location](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_intel_buffer_location.asciidoc).

Also, the code synopsis should not show the implementation details.  Just show  the API.  Therefore, just show something like:

```
class buffer_location {
 public:
  buffer_location(int location);
  int get_buffer_location() const;
};
```

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:71]: When you use the new document template, make sure the name of the feature-test macro is exactly the same as the name of the extension (but in upper case).

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (sherry-yuan) [reply sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc]: Thanks for the pointer to template!

Currently this extension will be in intel extension (let me know if it makes more sense to be an oneapi extension

It is also an experimental extension given this is a workaround.

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (sherry-yuan) [reply sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:61]: Thanks I mentioned it in the [Note] section

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (sherry-yuan) [reply sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc]: Thanks! Namespace is modified to `sycl::ext::intel::experimental::property::usm`

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (sherry-yuan) [reply sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc]: Thanks for the example! Spec changed accordingly.

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (sherry-yuan) [reply sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc]: The feature test macro is now changed to `SYCL_EXT_INTEL_RUNTIME_BUFFER_LOCATION` to align with `sycl_ext_intel_runtime_buffer_location`

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:50]: I don't think this should be a NOTE in the status section.  We should use notes like this only when the current implementation is incomplete, and the note explain what is not yet  implemented.  In this case, the formal specification of the property should state that the property is ignored for non-FPGA devices.  That is the intended behavior of the property, not something that is missing due to an incomplete specification.

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:52]: I don't think you need to say this either.  Therefore, I think you can remove this NOTE entirely.

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:61]: I'd suggest removing this last sentence.  This is only the overview section.  There should be a more formal definition of the property below in the "Specification" section.

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:96]:
```suggestion
array = (int *)sycl::malloc_device<int>(
        N, q,
        sycl::property_list{sycl::ext::intel::experimental::property::usm::buffer_location(2)});

sycl::queue q;
q.parallel_for(sycl::range<1>(N), [=] (sycl::id<1> i){
  data[i] *= 2;
}).wait();
```

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:102]: Be more precise here.  Is the new property allowed only for `sycl::malloc_device()`, or is it allowed also for the other USM malloc's?  For example, say something like:

> This extension adds the new property `sycl::ext::intel::experimental::property::usm::buffer_location` which applications can pass in the `property_list` parameter to all overloads of the `sycl::malloc_device()` function.  Following is a synopsis of this property.

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:116]: Add a formal description of the property here.  This description should explain what the `location` parameter means.  This is also the place to explain what happens if you specify the property for a non-FPGA device.  I was suggesting earlier that it could simply be ignored for non-FPGA devices, but I suppose we could instead say that `sycl::malloc_device()` throws an exception in this case.  I'm not sure which would be more convenient for users.  This section of the document should tell users what to expect in this scenario, though.

BTW, are there any special restrictions about how the pointer is used inside the kernel?  If there are, those need to be documented here also.

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (sherry-yuan) [reply sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc]: Technically yes, other malloc api can also accept this property, however they will not have any functional effect. I have made this comment in the spec. It is currently phrased as:

```
This extension adds the new property `sycl::ext::intel::experimental::property::usm::buffer_location`` which applications can pass in the property_list parameter to all overloads of the `sycl::malloc_device()`, `sycl::malloc_shared()`, `sycl::malloc_host()`  function. However, this property has no effect when passed to `sycl::malloc_shared()`, `sycl::malloc_host()`. Following is a synopsis of this property.
```

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:94]:
```suggestion
This extension adds the new property `sycl::ext::intel::experimental::property::usm::buffer_location` which applications can pass in the property_list parameter to all overloads of the `sycl::malloc_device()`, `sycl::malloc_shared()`, and `sycl::malloc_host()`  functions. However, this property has no effect when passed to `sycl::malloc_shared()` or `sycl::malloc_host()`. Following is a synopsis of this property.
```

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:53]: Sorry, I have one more formatting comment.  Can you break all these long lines at 80 columns, matching the other text in the template?

**Review by gmlueck — COMMENTED** [review]
This is looking a lot better.  Thanks for all the changes.  I just have a couple other comments.

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (sherry-yuan) [reply sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc:109]: Yes, I think this is the conventional terminology used for describing FPGA global memories. The same terminology was also used in the accessor buffer location spec  `SYCL_INTEL_buffer_location`

File: sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc
Comment (sherry-yuan) [reply sycl/doc/extensions/experimental/sycl_ext_intel_runtime_buffer_location.asciidoc]: Thank you for all the quick reviews and thanks for making this process smooth for the workaround! I have break all lines to be less than 80 char.

**Review by gmlueck — DISMISSED** [review]

**Comment by gmlueck** [comment]
Oh, I just noticed that this file is added to the "experimental" directory.  Is this implemented yet?  I think it is not?

If it is not implemented, the file should be added to "proposed".  It can then be moved to "experimental" in the PR that adds the implementation.  Or, if the implementation is simple, you can add the implementation to this PR and use this PR for both the specification and the implementation.

**Comment by sherry-yuan** [comment]
> Oh, I just noticed that this file is added to the "experimental" directory. Is this implemented yet? I think it is not?

Thanks, I have moved the spec to proposed folder. The implementation is already in another PR, I will keep it there for now.

**Review by gmlueck — APPROVED** [review]

**Comment by sherry-yuan** [comment]
I do not have write access to the repo, it will be nice if someone could merge it for me once all the blocking steps are done (if any).

**Comment by gmlueck** [comment]
> I currently do not have write access to the repo, it will be nice if someone could merge it for me once all the blocking steps are done (if any).

Most people do not have permission to merge PRs. @bader usually merges them after there is an approval and after testing completes.

**Comment by bader** [comment]
> > I currently do not have write access to the repo, it will be nice if someone could merge it for me once all the blocking steps are done (if any).
> 
> Most people do not have permission to merge PRs. @bader usually merges them after there is an approval and after testing completes.

Please, notify @intel/llvm-gatekeepers team when PR is ready for merge.

**Comment by sherry-yuan** [comment]
@intel/llvm-gatekeepers The PR is ready to merge.

**Pull request merged**
